CHANGELOG.md

```diff
@@ -42,6 +42,7 @@ The Product Changelog at **[matomo.org/changelog](https://matomo.org/changelog)*
       * For progress bars use the methods `initProgressBar`, `startProgressBar`, `advanceProgressBar` and `finishProgressBar`
       * Tables can be rendered using the new method `renderTable`
     * For executing another command within your command use the new method `runCommand`
+* Requests sent by Matomo to plugins.matomo.org will no longer include an `HTTP_X_FORWARDED_FOR` header containing the current user's IP address. If you use an outbound proxy rule that used this header to allow access for Matomo then it should be replaced with rule allowing access by IP and/or URL.    
```

Comment:
Actually that is the case for ALL outgoing requests using the HTTP class. This affects not only `plugins.matomo.org`, but also requests to download matomo, geo ip databases, and while thinking about this, I'm unsure if this might have a "bad" effect on our SEO plugin. That ones sends requests to google, bing,... Not sure if providing a `HTTP_X_FORWARDED_FOR` in that might not still be better, as otherwise the requests could be blocked as too many requests are coming from the same IP maybe.

Reply:
Good point @sgiehl :+1:

The removed code would only modify the `HTTP_X_FORWARDED_FOR` header if it was already set, for on-premise installations not accessed via a proxy or load balancer the header wouldn't be set on request to Matomo anyway so outgoing SEO plugin requests from Matomo wouldn't have the header set either. For most on-premise installations it probably isn't going to chance anything.

It could potentially be a problem for matomo.cloud or other services where many Matomo installations are using a limited pool of IP addresses for external requests and the use of load balancers means the `HTTP_X_FORWARDED_FOR` header would probably be set for all user connections therefore would be modified with the user IP for outbound connections too.

That said we don't really want to be leaking the admin user IP address to google or bing either if possible.

We could perhaps add an opt-in property to the HTTP class to add an `HTTP_X_FORWARDED_FOR` header with the user IP but only use it where necessary and allowed via config? So the SEO plugin or GeoIP download could use the header for requests to google, bing, etc but there could still be a global option to disable this for maximum privacy when it's not needed by the installation. 

Something like:

| Request could benefit from using header? |  Allowed via config?  | Result |
| --- | --- | --- |
| No | Yes | Don't use header |
| No | No | Don't use header |
|Yes | No | Don't use header |
| Yes | Yes | Use header |

Any thoughts on this one @samjf @tsteur ? We previously discussed this change in the context of potentially breaking outbound proxies, but not the possibility of causing third party requests to rate limit cloud so it might be worth revisiting this in an abundance of caution.

Reply:
@sgiehl since on the Cloud we're running core/plugins code it be great for the core/plugins teams to check if there could be any issues around rate limiting from the services that the core/plugins connect to.

Reply:
In Cloud specific code AFAIK we don't really connect to any other service that would cause any issue

Reply:
Sorry I missed this tagging. In general for cloud I don't think it should be a problem. We do centralised updates of vendor libraries like GeoIP so it shouldn't affect that. I believe that the Google APIs we are interested in have the ability to assign arbitrary users for a request to divide quotas and handle that rate limiting problem.
I would suspect (from what I understand) that most places wouldn't use X_FORWARDED_FOR for rate limiting anyway as they can't establish a trust relationship with an unknown proxy. So I would think it wouldn't harm much.

Reply:
Thanks @samf, sounds like this change will be fine as is :+1: